Echo and Replacer

ddiniz-m opened the pull request:
The branch is called heredoc, but no heredoc was worked on.

Instead I handled environment variables and echo.

The way I did it is substitute the strings in the main for the respective env variables. Handling quotes, $ and other annoying cases.

ddiniz-m committed:
Added necessary files

ddiniz-m committed:
transfered pipex files

ddiniz-m committed:
meh

ddiniz-m committed:
Makefile

ddiniz-m committed:
Added env_export functions. Worked on exec

ddiniz-m committed:
Exec function works. Deletd env and export files that I previously added because it became too confusing

ddiniz-m committed:
Updated files

ddiniz-m committed:
Exec now forks.

ddiniz-m committed:
Quick small change

ddiniz-m committed:
Started working on pipex related functions.
The only thing that works is wc -l > test (redir output). It has a small leak tho.
Redirect input works but it exits immediatly. (leaks idk)
fuck you pipes

ddiniz-m committed:
Small tests. All previous errors apply

ddiniz-m committed:
I think the leaks are resolved

ddiniz-m committed:
Getting things up to date

ddiniz-m committed:
>, <, >> works. << still doesnt (it works like < for now)
Started working on pipes.

ddiniz-m committed:
I'll have to rework redirects beacause it only deals with redirects of the first command:
ls -l > file.txt WORKS
ls -l > file.txt | wc -l < file.txt WOULD'T WORK (if pipes worked)

Pipes:
echo a | echo b WORKS;
ls -l | wc -l WON'T WORK (It doesn't return anything) (most probably dup2 issues);

Others:
sort by itself goes infinitely (its supposed to), but when CTRL-Cing, it creates another
minishell prompt and it doesnt let you CTRL-D

ddiniz-m committed:
This commit accounts for the changes I did yestarday, but beacuse I broke it so much
it didn't make sense to commit;

I just worked on pipes and redirects. Some things work, some half work and some just dont.
I made a txt file where I explain my thought process, for my own organization/unserstanding.

ddiniz-m committed:
Fixed mutiple redirects. For example ls -l > file1 > file2 now clears file1 and puts ls -l ouput
in file2.

Fixed an issue where the program would execute the last command twice, resulting in doubled outputs.

Half fixed program exiting when using pipe:
 - cat file | sort; would exit program.
 - ls -l > file1 | ls -l > file2; still exits.

Have to make more tests, especially with inputs (<).

Maybe will work on a script that'll compare results with actual bash.

ddiniz-m committed:
Implemeneted pipe process fix that Martim figured out.
Divided functions (for norminette)
Organized functions in respective files (Can be ajusted).

ddiniz-m committed:
Forgot to add files to Makefile and functions to .h
Also rearranged files and functions again.

Known errors:
When executing a command that should have some kind of input (wc, sort) it goes infinite
 (as is supposed to). The problem is when CTRL-Cing, it creates new line but it
 repeats the prompt:

	minishell$>sort
	^C
	minishell$>minishell$>

ddiniz-m committed:
Added our env, export and unset built-ins.
Had to change some exec/ functions beacause they would mainly use the envp from
main.
Basically removed all char **envp and started using our env that is in the
minishell struct.
Also had to make a list-to_array function to use in execve.

Results:
 Env and export by themselves work fine. When exporting variable it wont show
in either. No leaks.
 Unseetting a variable that came from main (export in bash, enter minishell
and unset) it leaks.

ddiniz-m committed:
Fixed export, env and unset.

Added functionality:
  - command file | export/env/unset

Fixed redirection and pipe issues:
  - cat file1 | wc -l (Would output 1);
  - < file1 cat | file1 < wc -l > file2 (cat would output to terminal);
  - ls -l > file1 | ls -l > file2 (last command would output to terminal);

Fixed issue that occured in ft_strnstr when trying to get PATH str from env list.

Known errors:
If a command is running and CTRL-C is pressed, it repeats the prompt. (E.g. wc)

ddiniz-m committed:
Fixed:
When running a non command/built-in (e.g. 'hello'), exec would return and not exit resulting in multiple open childs.

ddiniz-m committed:
Small fix

ddiniz-m committed:
Co-authored-by: WudDoo <WudDoo@users.noreply.github.com>

ddiniz-m committed:
Implemented environment variables. (e.g. $HOME)
 - When, for example, $HOME is written, it edits the main_arrays with the value.

Added cd and echo.
 - Edited echo so it works with the new structs/lists - Works fine.
 - Edited cd so it uses our env.
 - cd doesn't work.

Exec branch could now be the buffer where we test shit before merging with main.

ddiniz-m committed:
heredoc

mortinso committed:
Fixed exit
exit and cd don't work, but that's an exec error.
I'll keep trying to fix exec, and I'll check for leaks.

ddiniz-m committed:
Fixed leaks from exec. Ended up not advancing heredoc. Helped Dinis figure out philosophers lol

mortinso committed:
Fixed a leak, etc.
I can't find any other leaks.
exit and cd still don't work because of exec.
Pipes aren't working properly but that may be because I'm using codespaces.
Pipe tests that didn't work:
    - ```env | grep <variable>```
    - ```export | wc -l ```
I have to stop for today but I'll try to fix this morning.

mortinso committed:
Massive exec fix
Changes to be committed:
modified:   inc/minishell.h
- changed cd() prototype

modified:   src/built-ins/cd.c
- now gives a proper error message when it receives multiple arguments

modified:   src/exec/exec_utils.c
- fixed leaks when using `exit` inside pipeline
- `exit` now works properly when it's the last command of a pipeline

modified:   src/exec/open_file.c
- added proper error messages instead of "OPEN ERROR"

modified:   src/exec/run_pipes.c
- `exit` now works properly when it's the only command called
- minor redundancy clean up on run()

modified:   src/main.c
- removed unnecessary comments

modified:   src/structs/content.c
- fixed dumb mistake that was causing errors on `cd | ls -l`

mortinso committed:
Commented some stuff that's wrong

ddiniz-m committed:
Merge branch 'exec_copy' into heredoc

ddiniz-m committed:
I though this was gonna be simple...

Branch up to date with exec_copy.
Added GNL.
Regarding hdoc, here's what works (or at least what I tested):
 - Simple hdoc with no commands (<< EOF)
 - hdoc with commands (echo a << EOF)
 - Multiple hdocs (<< hello << EOF)

Here's what doesn't
 - CTRL-C in the middle of the hdoc does the double prompt thing and it doesn't exit cleanly
 - History doesn't work properly. Only the main str is added. (Should add main->str, hdoc contents and
   delimiter.

What I was in the middle of doing (and thinking):
 - I figured we are going to have to save all hdoc input in a temporary file and then strjoin
   it all together to put in add_histroy(). I created a fd_int and opened it in init.c/var_init().
   For now I just ft_putstr_fd'd the correct strings to tmp_file.
 - I need to test hdoc with pipes. Dont remember what (<< EOF | << EOF) does
 - CTRL-C seems to be the same error that occurs when interrupting child while running (wc and sort).
 - More tests of course.

ddiniz-m committed:
Fixed strncmp to compare until strlen of line - 1 (because of \n);

Removed changes I made to deal with add_history because its not needed.

ddiniz-m committed:
Fixed export and env.
Added echo functionalities that were missing. (incomplete)
Commented str_utils funcions.
Fixed strcmp_chr beacuse it was bad.

mortinso committed:
exit|exit still doesnt work brh moment ong ong

ddiniz-m committed:
Fixed echo dealing with quotes

ddiniz-m committed:
Merge branch 'main' into heredoc

ddiniz-m committed:
Fixed Echo. Merged with main, so its up to date. Fixed issues whem PATH is unset. Missing error when trying to run something that is not an exec/built-in.

mortinso committed:
Finished?

Removed our attempt at redirecting stderr to a `pipe_fd`

Pipeline behavior fixes:
- `exit` no longer writes "exit\n" to stderr when in a pipeline
- `cd` no longer changes directories when in a pipeline

`g_exit`:
- I think I fixed `g_exit`'s behavior in `exec()`, need more testing
- Not sure if the child processes are changing g_exit, will need further testing

`last_cmd()`:
- Changed function to `void` type from `int` type, since it was never returning anything other than zero
- Removed a redundancy I had made

mortinso committed:
organized files

mortinso committed:
some norminette compliance and organization

mortinso committed:
norminette compliance (except some comments)

ddiniz-m committed:
Merge remote-tracking branch 'martim-branch/WudDoo-exit_cd_fixpipes' into heredoc

ddiniz-m committed:
Forgot one modification

ddiniz-m committed:
Worked on substituting any occurence of $ by its env value.

The only problem I see is when doing "$HOME"aaaa it does't work.

In quote_var is the space to put a function to deal with that.

mortinso committed:
merge main into heredoc

mortinso committed:
fixed an error on declare -x BROWSER="/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc/bin/helpers/browser.sh"
declare -x CLOUDENV_ENVIRONMENT_ID="8ec0ed24-58ae-4bf6-b296-27d094468f2a"
declare -x CODESPACES="true"
declare -x CODESPACE_NAME="friendly-guacamole-5p5q5965x9jfv5gp"
declare -x CODESPACE_VSCODE_FOLDER="/workspaces/minishell"
declare -x COLORTERM="truecolor"
declare -x CONDA_DIR="/opt/conda"
declare -x CONDA_SCRIPT="/opt/conda/etc/profile.d/conda.sh"
declare -x ContainerVersion="13"
declare -x DEBIAN_FLAVOR="focal-scm"
declare -x DOCKER_BUILDKIT="1"
declare -x DOTNET_ROOT="/usr/local/dotnet/current"
declare -x DOTNET_SKIP_FIRST_TIME_EXPERIENCE="1"
declare -x DYNAMIC_INSTALL_ROOT_DIR="/opt"
declare -x ENABLE_DYNAMIC_INSTALL="true"
declare -x GEM_HOME="/usr/local/rvm/gems/ruby-3.2.2"
declare -x GEM_PATH="/usr/local/rvm/gems/ruby-3.2.2:/usr/local/rvm/gems/ruby-3.2.2@global"
declare -x GITHUB_API_URL="https://api.github.com"
declare -x GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN="app.github.dev"
declare -x GITHUB_CODESPACE_TOKEN="AVM2TW3E7EFMB4MWCOQWHBDFFYEG7ANCNFSM4AFGW6YA"
declare -x GITHUB_GRAPHQL_URL="https://api.github.com/graphql"
declare -x GITHUB_REPOSITORY="ddiniz-m/minishell"
declare -x GITHUB_SERVER_URL="https://github.com"
declare -x GITHUB_TOKEN="ghu_zw5OclomLQrsnRnUpK09AQQ4iwVMah3O4cV6"
declare -x GITHUB_USER="WudDoo"
declare -x GIT_ASKPASS="/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc/extensions/git/dist/askpass.sh"
declare -x GIT_EDITOR="code --wait"
declare -x GOPATH="/go"
declare -x GOROOT="/usr/local/go"
declare -x GRADLE_HOME="/usr/local/sdkman/candidates/gradle/current"
declare -x HOME="/home/codespace"
declare -x HOSTNAME="codespaces-2cad7b"
declare -x HUGO_DIR="/usr/local/hugo/bin"
declare -x HUGO_ROOT="/home/codespace/.hugo"
declare -x INTERNAL_VSCS_TARGET_URL="https://uksouth.online.visualstudio.com"
declare -x IRBRC="/usr/local/rvm/rubies/ruby-3.2.2/.irbrc"
declare -x JAVA_HOME="/usr/local/sdkman/candidates/java/current"
declare -x JAVA_ROOT="/home/codespace/java"
declare -x JUPYTERLAB_PATH="/home/codespace/.local/bin"
declare -x LANG="C.UTF-8"
declare -x LESSCLOSE="/usr/bin/lesspipe %s %s"
declare -x LESSOPEN="| /usr/bin/lesspipe %s"
declare -x LS_COLORS="rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:"
declare -x MAVEN_HOME="/usr/local/sdkman/candidates/maven/current"
declare -x MAVEN_ROOT="/home/codespace/.maven"
declare -x MY_RUBY_HOME="/usr/local/rvm/rubies/ruby-3.2.2"
declare -x NODE_ROOT="/home/codespace/nvm"
declare -x NPM_GLOBAL="/home/codespace/.npm-global"
declare -x NUGET_XMLDOC_MODE="skip"
declare -x NVM_BIN="/usr/local/share/nvm/versions/node/v20.8.0/bin"
declare -x NVM_CD_FLAGS=""
declare -x NVM_DIR="/usr/local/share/nvm"
declare -x NVM_INC="/usr/local/share/nvm/versions/node/v20.8.0/include/node"
declare -x NVM_SYMLINK_CURRENT="true"
declare -x NVS_DIR="/usr/local/nvs"
declare -x NVS_HOME="/usr/local/nvs"
declare -x NVS_OS="linux"
declare -x NVS_ROOT="/usr/local/nvs"
declare -x NVS_USE_XZ="1"
declare -x OLDPWD="/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc"
declare -x ORYX_DIR="/usr/local/oryx"
declare -x ORYX_ENV_TYPE="vsonline-present"
declare -x ORYX_PREFER_USER_INSTALLED_SDKS="true"
declare -x ORYX_SDK_STORAGE_BASE_URL="https://oryx-cdn.microsoft.io"
declare -x PATH="/usr/local/rvm/gems/ruby-3.2.2/bin:/usr/local/rvm/gems/ruby-3.2.2@global/bin:/usr/local/rvm/rubies/ruby-3.2.2/bin:/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc/bin/remote-cli:/home/codespace/.local/bin:/home/codespace/.dotnet:/home/codespace/nvm/current/bin:/home/codespace/.php/current/bin:/home/codespace/.python/current/bin:/home/codespace/java/current/bin:/home/codespace/.ruby/current/bin:/home/codespace/.local/bin:/usr/local/python/current/bin:/usr/local/py-utils/bin:/usr/local/oryx:/usr/local/go/bin:/go/bin:/usr/local/sdkman/bin:/usr/local/sdkman/candidates/java/current/bin:/usr/local/sdkman/candidates/gradle/current/bin:/usr/local/sdkman/candidates/maven/current/bin:/usr/local/sdkman/candidates/ant/current/bin:/usr/local/rvm/gems/default/bin:/usr/local/rvm/gems/default@global/bin:/usr/local/rvm/rubies/default/bin:/usr/local/share/rbenv/bin:/usr/local/php/current/bin:/opt/conda/bin:/usr/local/nvs:/usr/local/share/nvm/versions/node/v20.8.0/bin:/usr/local/hugo/bin:/usr/local/dotnet/current:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/rvm/bin"
declare -x PHP_PATH="/usr/local/php/current"
declare -x PHP_ROOT="/home/codespace/.php"
declare -x PIPX_BIN_DIR="/usr/local/py-utils/bin"
declare -x PIPX_HOME="/usr/local/py-utils"
declare -x PROMPT_DIRTRIM="4"
declare -x PWD="/workspaces/minishell"
declare -x PYTHONIOENCODING="UTF-8"
declare -x PYTHON_PATH="/usr/local/python/current"
declare -x PYTHON_ROOT="/home/codespace/.python"
declare -x RAILS_DEVELOPMENT_HOSTS=".githubpreview.dev,.preview.app.github.dev,.app.github.dev"
declare -x RUBY_HOME="/usr/local/rvm/rubies/default"
declare -x RUBY_ROOT="/home/codespace/.ruby"
declare -x RUBY_VERSION="ruby-3.2.2"
declare -x RVM_PATH="/usr/local/rvm"
declare -x SDKMAN_CANDIDATES_API="https://api.sdkman.io/2"
declare -x SDKMAN_CANDIDATES_DIR="/usr/local/sdkman/candidates"
declare -x SDKMAN_DIR="/usr/local/sdkman"
declare -x SDKMAN_PLATFORM="linuxx64"
declare -x SHELL="/bin/bash"
declare -x SHELL_LOGGED_IN="true"
declare -x SHLVL="2"
declare -x TERM="xterm-256color"
declare -x TERM_PROGRAM="vscode"
declare -x TERM_PROGRAM_VERSION="1.83.1"
declare -x USER="codespace"
declare -x VSCODE_GIT_ASKPASS_EXTRA_ARGS=""
declare -x VSCODE_GIT_ASKPASS_MAIN="/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc/extensions/git/dist/askpass-main.js"
declare -x VSCODE_GIT_ASKPASS_NODE="/vscode/bin/linux-x64/f1b07bd25dfad64b0167beb15359ae573aecd2cc/node"
declare -x VSCODE_GIT_IPC_HANDLE="/tmp/vscode-git-78e820a093.sock"
declare -x VSCODE_IPC_HOOK_CLI="/tmp/vscode-ipc-289d3ead-dc60-42cf-87b2-cfa0ae52f463.sock"
declare -x rvm_alias_expanded
declare -x rvm_bin_flag
declare -x rvm_bin_path="/usr/local/rvm/bin"
declare -x rvm_delete_flag
declare -x rvm_docs_type
declare -x rvm_file_name
declare -x rvm_gemstone_package_file
declare -x rvm_gemstone_url
declare -x rvm_hook
declare -x rvm_niceness
declare -x rvm_nightly_flag
declare -x rvm_only_path_flag
declare -x rvm_path="/usr/local/rvm"
declare -x rvm_prefix="/usr/local"
declare -x rvm_pretty_print_flag
declare -x rvm_proxy
declare -x rvm_quiet_flag
declare -x rvm_ruby_alias
declare -x rvm_ruby_bits
declare -x rvm_ruby_configure
declare -x rvm_ruby_file
declare -x rvm_ruby_global_gems_path
declare -x rvm_ruby_make
declare -x rvm_ruby_make_install
declare -x rvm_ruby_mode
declare -x rvm_ruby_string
declare -x rvm_ruby_url
declare -x rvm_script_name
declare -x rvm_sdk
declare -x rvm_shell_nounset
declare -x rvm_silent_flag
declare -x rvm_sticky_flag
declare -x rvm_system_flag
declare -x rvm_use_flag
declare -x rvm_user_flag
declare -x rvm_version="1.29.12 (latest)" where it was creating new variables on env when it should be overriding

mortinso committed:
Last commit name was fucked.

It was supposed to say:
"fixed an error on `export` where it was creating new variables on `env` when it should be overriding"
For some reason it replaced `export` by it's command output, whatever.

mortinso committed:
`is_exec()` now writes error message to stderr

ddiniz-m committed:
Did a massive overhaul to env_utl.c .Now for every str with an $, it splits it into and array.
For example str = a$HOME"$HOME"$PATH
becomes
arr0 a
arr1 $HOME
arr2 "$HOME"
arr3 $PATH

So I can handle each individually.

ddiniz-m committed:
Forgot to add:
The only thing(i think) that doesnt work is '$HOME' doesnt print anything.
Some functions still have +25 lines and env_utl.c has +5 funcs.

ddiniz-m committed:
Fixed most errors regarding env_str.

Still need to tidy up env_utl.c

Don't know why but some previously working features dont worka anymore (Problems with cd and running minishell inside minishell)

ddiniz-m committed:
Cleaned up env_utl.c by creating more files.

Fixed (echo "$HOME a") returning NULL instead of (/nfs/home a)

When doing (echo "'$HOME'") it should print ('/nfs/home/'). Instead it just prints ('$HOME')

Have to go, so I left a small leak when doing the above prompt.

ddiniz-m committed:
a

ddiniz-m committed:
Env variables are working.

Norm errors are fixed

ddiniz-m committed:
Merge branch 'main' into heredoc

mortinso commented:
Our minishell doesn't interpret `echo "'$HOME'$PWD"` correctly
![image](https://github.com/ddiniz-m/minishell/assets/89762267/83ec56f9-814d-49d4-a53f-8bc08fb0baa0)
![image](https://github.com/ddiniz-m/minishell/assets/89762267/5846fd6e-63d3-4817-9a4f-45947f2685b0)

mortinso commented:
It also doesn't handle `echo $HOME_m` correctly
![image](https://github.com/ddiniz-m/minishell/assets/89762267/ae311660-bb1b-428f-bc3f-0a078465ff2d)
![image](https://github.com/ddiniz-m/minishell/assets/89762267/2cf227ba-3fc3-40d9-9f06-d224793b5f26)

mortinso commented:
`echo $HOME1` is also incorrect
![image](https://github.com/ddiniz-m/minishell/assets/89762267/9a195d6b-e05d-42a8-aa5c-295513ae4e7b)
![image](https://github.com/ddiniz-m/minishell/assets/89762267/324c717a-f27d-4ebb-b305-8f5033452797)

mortinso commented:
Notes:
- `_` can be part of an environment value
- numbers can be part of an environment value
- single-quotes should be printed when inside double-quotes

mortinso commented:
Our minishell SEGV's on `echo "$HOME'a"`
![image](https://github.com/ddiniz-m/minishell/assets/89762267/9b55fb62-3612-45a5-aa64-b4d2cf6eac71)
![image](https://github.com/ddiniz-m/minishell/assets/89762267/d0b71e14-b7b7-4eea-839f-b256b42035c4)

mortinso committed:
Dumb Makefile mistake

mortinso committed:
Partway there with the replacer

Right now fails the following situations:
- `$VAR"a"`
- `"'$VAR'"`
- `"$VAR'a"`
- `$"VAR"`
- `$'VAR'`

We should keep testing to make sure its working in all cases. I added a text file with the cases I've been testing for.
I was able to fix 5 cases today, the other shouldn't be too much trouble to fix.

mortinso commented:
Need to fix the following situations:

- [x] `$VAR"a"`
- [x] `"'$VAR'"`
- [x] `"$VAR'a"`
- [x] `$"VAR"`
- [x] `$'VAR'`

We should keep testing to make sure its working in all cases. I added a text file with the cases I've been testing for.

mortinso commented:
We should probably make a separate folder for the replacer

ddiniz-m committed:
Fixed more cases of env_var.

ddiniz-m committed:
I think replacer is mostly done.

The only thing I just broke is single quotes printing the value of variables. Must be easy to fix.

ddiniz-m committed:
Reworked replace.

Current errors:
echo '$'HOME
echo $'HOME'

mortinso committed:
merge main into heredoc

mortinso committed:
merge main into heredoc

ddiniz-m committed:
Fixed all issues.

ddiniz-m commented:
Had errors. Will make a new one

ddiniz-m committed:
Norm

mortinso committed:
merge main into heredoc

mortinso committed:
fixed merge error

mortinso committed:
fixed another merge error

mortinso committed:
Update exit.c

Fixed `exit_format_error()`

mortinso committed:
Update syntax_errors2.c

Fixed an error

mortinso committed:
Update syntax_errors.c

Fixed errors

mortinso committed:
Update main.c

error fix

mortinso committed:
Update syntax_errors.c

error

mortinso committed:
Update cmd_utils.c

fixed an error

mortinso committed:
Delete tests and expected results.txt

mortinso committed:
fixed another error

mortinso committed:
fixed last error i could find

mortinso commented:
I think all the syntax error functions are fixed now. We should test tomorrow though.

mortinso committed:
Removed `skip_rquotes()` as it wasn't being used anywhere

mortinso committed:
fixed an error on cd and export

mortinso committed:
fixed part of the 'Binary file (standard input) matches' error

mortinso committed:
88/146

mortinso committed:
91/146

mortinso committed:
92/146

mortinso committed:
94/146

mortinso committed:
98/146

mortinso committed:
99/146 \n all built-ins are supposedly working properly and so are pipes

mortinso commented on src/built-ins/cd.c:
cd shouldn't exit

mortinso commented on src/run/redirections.c:
should actually create an open error function

mortinso commented on src/signals.c:
should it write exit?

mortinso committed:
`cd()` no longer exits 136/146

mortinso committed:
Between 138 and 140/146
